source/opt/log.h

```diff
@@ -0,0 +1,202 @@
+// SPIRV_ASSERT(<message-consumer>, <condition-expression>, <message>);
+// SPIRV_ASSERT(<message-consumer>, <condition-expression>,
+//              <message-format>,   <variable-arguments>);
+//
+// In the third format, the number of <variable-arguments> cannot exceed (5 -
+// 2). If more arguments are wanted, grow PP_ARG_N and PP_NARGS in the below.
+#if !defined(NDEBUG)
+#define SPIRV_ASSERT(consumer, ...) SPIRV_ASSERT_IMPL(consumer, __VA_ARGS__)
+#else
+#define SPIRV_ASSERT(consumer, ...)
+#endif
+
+// Logs a debug message to the consumer. Accepts the following formats:
+//
+// SPIRV_DEBUG(<message-consumer>, <message>);
+// SPIRV_DEBUG(<message-consumer>, <message-format>, <variable-arguments>);
+//
+// In the second format, the number of <variable-arguments> cannot exceed (5 -
+// 1). If more arguments are wanted, grow PP_ARG_N and PP_NARGS in the below.
+#if !defined(NDEBUG) && defined(SPIRV_LOG_DEBUG)
+#define SPIRV_DEBUG(consumer, ...) SPIRV_DEBUG_IMPL(consumer, __VA_ARGS__)
+#else
+#define SPIRV_DEBUG(consumer, ...)
+#endif
+
+// Logs an error message to the consumer saying the given feature is
+// unimplemented and abort the program.
+#define SPIRV_UNIMPLEMENTED(consumer, feature)                               \
+  do {                                                                       \
+    spvtools::Log(consumer, MessageLevel::Fatal, __FILE__, {__LINE__, 0, 0}, \
+                  "unimplemented: " feature);                                \
+  } while (0)
+
+// Logs an error message to the consumer saying the code location
+// should be unreachable and abort the program.
+#define SPIRV_UNREACHABLE(consumer)                                \
+  do {                                                             \
+    spvtools::Log(consumer, MessageLevel::InternalError, __FILE__, \
+                  {__LINE__, 0, 0}, "unreachable");                \
+  } while (0)
+
+// Helper macros for stringify the given argument.
+#define SPIRV_STRINGIFY(s) SPIRV_STRINGIFY_(s)
+#define SPIRV_STRINGIFY_(s) #s
+
+// Helper macros for concatenating arguments.
+#define SPIRV_CONCATENATE(a, b) SPIRV_CONCATENATE_(a, b)
+#define SPIRV_CONCATENATE_(a, b) a##b
+
+// Helper macro to force expanding __VA_ARGS__ to satisfy MSVC compiler.
+#define PP_EXPAND(x) x
+
+namespace spvtools {
+
+enum { INIT_BUFFER_SIZE = 256 };
+
+// Calls the given |consumer| by supplying  the |message|. The |message| is from
+// the given |file| and |location| and of the given severity |level|.
+inline void Log(const MessageConsumer& consumer, MessageLevel level,
+                const char* file, const spv_position_t& position,
+                const char* message) {
+  if (consumer != nullptr) consumer(level, file, position, message);
+
+  // TODO(antiagainst): This could tear down the whole problem if called in a
+  // library. But do we have better solution?
+  if (level == MessageLevel::Fatal || level == MessageLevel::InternalError)
+    exit(EXIT_FAILURE);
```

Comment:
Note: You can't actually test this case (except through weird weird mechanisms.  So that's a smell.

At least two alternatives:
 - accept an "exit" function pointer that is called in this case.  That allows you to inject the dependency.   But it would be something like "throw an exception" which is ugly, or setjmp/longjmp which is worse in C++.
 - don't exit!   I like this alternative.   For example, maybe Log should return the error classification enum, and then the caller can choose to do what they want.  You can then wrap common behaviours as you like.